### ISSUE TITLE
Pushing 20k items into a leaf-list in the operational DS

### ISSUE THREAD

**Issue by jktjkt** [sysrepo/sysrepo#2288]
Hi, following up from our call on Friday, I'm trying to push ops data, updating a 20'000-items `leaf-list` of `decimal64` in one go. I've got two questions so far:

### 1. How do I create the corresponding `lyd_data` data structure efficiently?

I've tried code like this:
```cpp
for (const auto & [path, value] : data) {
    if (!out) {
        out = std::make_shared<libyang::Data_Node>(session->get_context(),
                path.c_str(), utils::toString(value).c_str(),
                LYD_ANYDATA_CONSTSTRING, LYD_PATH_OPT_OUTPUT);
    } else {
        out->new_path(session->get_context(),
                path.c_str(), utils::toString(value).c_str(),
                LYD_ANYDATA_CONSTSTRING, LYD_PATH_OPT_OUTPUT);
    }
}
```
...which, if I'm not mistaken, corresponds to something like this:
```c
struct lyd_node* node = NULL;
for (int i = 0; i < 20000; ++i) {
  node = lyd_new_path(node, ctx, pathForMePlease(i), nullptr, LYD_ANYDATA_CONSTSTRING, LYD_PATH_OPT_OUTPUT);
}
```
I measured both the C and C++ code, they're comparable. Constructing this `lyd_node` hierarchy directly in C takes about 2.3s on my laptop from 2019, and that's just libyang, before I even push stuff to sysrepo. End-to-end with just one subscriber (whose performance is effectively just noise compared to the initial work of preparing 20k `lyd_node`s), it takes about 3.4s for a one-time push of 20k items into the ops datastore. Here's the flame graph from the profiler:
![image](https://user-images.githubusercontent.com/2631925/104857130-ac4fde00-5916-11eb-9b45-6008dff2969d.png)

I think I must be doing something stupid, so how do I push 20k items into a `leaf-list`? We can assume that I'm always providing a new `leaf-list`. Which goes to...

### 2. How do I replace all items of a `leaf-list` in the first place?

I think this is that dreaded `merge` operation from NETCONF. I'm using a `leaf-list` to represent a spectral scan. I'm always pushing a new "vector of numbers", the previous versions have no meaning. This NETCONF convention and the associated behavior of treating nodes with the same XPath as "the same nodes" is causing me headache :).

Let's say I've pushed a `leaf-list` that holds three numbers, `{1, 2, 3}`. How do I update it to `{2, 1, 5}` in the sysrepo `operational` DS?

**Comment by jktjkt** [comment]
The XPath encoding made me double-check if duplicate items in a `leaf-list` are allowed in the first place. I did not find much in the YANG 1.1 RFC, but the RESTCONF one [says](https://tools.ietf.org/html/rfc8040#section-3.5.3) that:
> - YANG 1.1 allows duplicate leaf-list values for non-configuration data.  In this case, there is no mechanism to specify the exact matching leaf-list instance.

On the other hand, [pyang apparently enforces no duplicates](https://github.com/robshakir/pyangbind/issues/179).

Just a random datapoint here, feel free to tell me that `leaf-list` is not currently that awesomely supported, and that sticking with a full-blown `list` is a better way.

**Comment by jktjkt** [comment]
Wow, with `list` (and once again 20k items), this ops data push takes about 0.7s per each iteration. That's much better (almost awesome, now awesome would be <0.1s). Let's hope that it doesn't take much longer on our ARM boards. Rebuilding libyang with `-DENABLE_CACHE=ON` doesn't seem to improve this.

I wanted to avoid allocating these 60k `lyd_data` nodes (20k list items * 3 nodes per each (one node per "list item" + one node for the key + one node for the measured value)) over and over again -- i.e., to reuse them between ops pushes. That had a really horrible performance -- it did not finish in a reasonable time with 20k items, while doing "something useful" for 333 items -- I suppose `lyd_find_path` over a list is O(n) and `lyd_new_path` is perhaps O(log N)?:

```cpp
for (const auto& [k, v] : data) {
    lyd_change_leaf(reinterpret_cast<struct lyd_node_leaf_list *>((lyd_find_path(edit->C_lyd_node(), k.c_str())->set.d)[0]),
       std::get<std::string>(v).c_str());
}
```

Should I try and iterate over that tree in some better way? Given a YANG tree like this:
```
  +--ro scan-structured-global
  |  +--ro spectrum* [frequency]
  |     +--ro frequency    frequency
  |     +--ro power?       power-dBm
```
Is it possible to iterate over all `power` leafs in the order of increasing `frequency`? These frequencies are static, once first set, they won't ever change.

Here's the picture from the profiler BTW. On a high-level look it seems to me that `sr_apply_changes` reads module data from SHM twice, consuming about 17% of CPU cycles. Is there something that strikes you as an optimization possibility?

![image](https://user-images.githubusercontent.com/2631925/104860356-66e9db80-592b-11eb-9a71-551824adf21f.png)

**Comment by michalvasko** [comment]
Answering chronologically.

1. Okay, the graph does not look that bad. First and foremost, do not use `lyd_new_path()` for this simple use-case but rather `lyd_new_leaf()`, that should completely get rid of the first bigger half of the graph (depending on the exact module, you may need to create its parents first using `lyd_new()`). Sadly, I cannot help you with the second half because the inefficiency of *libyang* insert functions was fixed by *libyang* version 2 and *sysrepo* using it will not be ready in time for your paper.
2. Well, there are several options here and I am not able to say which is the most efficient (possibly it is all the same). But the simplest should be simply using `replace` as the default operation of `sr_edit_batch()` (but this again depends on the exact module, would not work if the leaf-lists are top-level).

You are right about duplicates being allowed in state data, that is how *sysrepo* behaves. Also, `list` and `leaf-list` support should be the same.

I do not understand why `list` should have better performance than `leaf-list`. It is also interesting that the cache makes no difference but it is possible its optimizations are not used in this use-case.

> I wanted to avoid allocating these 60k lyd_data nodes

As you see, it may be the most efficient way after all.

> Is it possible to iterate over all power leafs in the order of increasing frequency? These frequencies are static, once first set, they won't ever change.

Yes, in that case you should be able to get the specific `spectrum` instances efficiently with `lyd_find_sibling_val()` (provided you use the cache). However, I am not sure what exactly you want to use this for because you can always simply create all the instances in this order and it will be kept so then a simple iteration over siblings should get you what you need.

> Here's the picture from the profiler BTW. On a high-level look it seems to me that sr_apply_changes reads module data from SHM twice, consuming about 17% of CPU cycles. Is there something that strikes you as an optimization possibility?

The data are needed twice (and whether you parse them or duplicate them should not make much difference) because first the push oper data are being constructed for the callbacks and then they are being updated and stored. This explanation may not be clear but it simply is needed.

So, as the first thing you should get rid of `lyd_new_path()`, should be quite simple. If that is not enough, i am not sure there is much I can do. You can try using *libyang* and *sysrepo* from `libyang2` branches but, while should generally be working (all the tests are passing), there may easily be many bugs and actually no optimizations were performed using real use-cases yet (so the final performance may even be worse). Also, *netopeer2* will not be available for that *sysrepo* that soon.

**Comment by jktjkt** [comment]
Let's start with something trivial, like a `leaf-list` of ints. Could you please give me an example on how to:

1. push in there `{1, 2, 3, 2}`,
2. get a CB notifying me "about a change" while being able to fetch the full `leaf-list`,
3. somehow update the content of that `leaf-list` in the ops DS to `{4, 5, 2, 1, 2}`
4. getting a CB where I can see that that is the new content

I tried that, and I failed horribly -- it seems that the modification CB just won't get me there. The modifier CB gives me access to a diff, which is absolutely not what I want. On the first invocation, I can get full ops DS content via something like:
```cpp
session->get_data("/dummy-scan:scan-flat-global")->print_mem(LYD_JSON, LYP_FORMAT)
```
...but that works only the first time. On a next change CB invocation, it prints out empty content. This makes me believe that it is not actually possible to work with ops leaf-list containing duplicate data in the ops DS.

Also, I've tried various version of `lyd_new`, `lyd_new_path` and `lyd_new_leaf`, but I keep getting error message about not being able to find a node, e.g., `Failed to find "scan" as a sibling to "dummy-scan:structured-scan".`. I suspect that the error message is wrong because I'm calling a function which takes a `parent`, so I expect that it should be looking for a **child**, not a **sibling**, right?

Here's the tree of my model:
```
module: dummy-scan
  +--ro scan-flat-global
  |  +--ro spectrum*            power-dBm
  |  +--ro lowest-freq?         frequency
  |  +--ro highest-frequency?   frequency
  +--ro scan-structured-global
     +--ro spectrum* [frequency]
        +--ro frequency    frequency
        +--ro power?       power-dBm
```

Can you help me how do I use `lyd_new_leaf`, please?

**Comment by michalvasko** [comment]
Okay, as it turns out, it was not supported. So, I have quickly added the support with a test (which you can take a look at to see how it works) and you must use `SR_EDIT_STRICT` if you want to create a duplicity. But like I said, I simply made the use-case in the test work, how many things can now be broken I have no idea. It may be fine for your use-case, though.

> I suspect that the error message is wrong because I'm calling a function which takes a parent, so I expect that it should be looking for a child, not a sibling, right?

The message is generic to work even for top-level nodes so it should be correct. If you provide a parent, you should see the first child of the parent in the message because its sibling is being searched for.

**Comment by jktjkt** [comment]
I went back to testing with a `leaf-list`. If I always provide the leaf-list values in a lexicographical order, it seems to work. However, when I either add `ordered-by user`, or when I start replacing these leaf-lists where data are provided in a different order, I'm getting an error from my `module_change_tree` CB:

```
libyang[0]: Invalid arguments - state leaf-list (lyd_find_sibling_val()).
```

Here's how I'm triggering this. Note that my "top level operation" is `merge` (or what do I use here? I want to make sure that my container that encloses the `lef-list` gets replaced, but the resit of the datastore should remain as-is), and I have a `replace` for the container that is the parent of my `leaf-list`.

```c++
void doEdit(std::shared_ptr<sysrepo::Session> session, const std::vector<std::string>& data) {
    auto edit = std::make_shared<libyang::Data_Node>(session->get_context(),
        (modulePrefix + "scan-flat-global").c_str(), "", LYD_ANYDATA_CONSTSTRING, LYD_PATH_OPT_OUTPUT);
    auto netconf = session->get_context()->get_module("ietf-netconf");
    edit->insert_attr(netconf, "operation", "replace");
    for (const auto& k : data) {
	edit->new_path(session->get_context(), k.c_str(), "", LYD_ANYDATA_CONSTSTRING, LYD_PATH_OPT_OUTPUT);
    }
    spdlog::get("main")->info("edit JSON: {}", edit->print_mem(LYD_JSON, 0));
    session->edit_batch(edit, "merge");
    session->apply_changes();
};

{
    std::vector<std::string> data{
        modulePrefix + "scan-flat-global/spectrum[.=\"1\"]",
        modulePrefix + "scan-flat-global/spectrum[.=\"2\"]",
        modulePrefix + "scan-flat-global/spectrum[.=\"3\"]",
    };
    doEdit(session, data);
}

{
    std::vector<std::string> data{
        modulePrefix + "scan-flat-global/spectrum[.=\"10\"]",
        modulePrefix + "scan-flat-global/spectrum[.=\"666\"]",
        modulePrefix + "scan-flat-global/spectrum[.=\"333\"]",
    };
    doEdit(session, data);
}

{
    std::vector<std::string> data{
        modulePrefix + "scan-flat-global/spectrum[.=\"1\"]",
        modulePrefix + "scan-flat-global/spectrum[.=\"10\"]",
        modulePrefix + "scan-flat-global/spectrum[.=\"333\"]",
        modulePrefix + "scan-flat-global/spectrum[.=\"666\"]",
        modulePrefix + "scan-flat-global/spectrum[.=\"-15.47\"]",
    };
    doEdit(session, data);
}
```

**Comment by jktjkt** [comment]
I made a mistake -- `Invalid arguments - state leaf-list (lyd_find_sibling_val()).` is only reported for `ordered-by user`. If I use the system order and provide my data in an arbitrary order, I do not get that error. Sorry for noise.

**Comment by michalvasko** [comment]
Right, that is exactly what I meant when I said that the support for push operational data is just a quick fix rather than a proper implementation. But as long as you will not have duplicate values, it should well with the provided patch (for *libyang*). If you do, it should work as well, but the results may not be as expected and the order may be wrong.

[find_sibling_state.patch.txt](https://github.com/sysrepo/sysrepo/files/6092012/find_sibling_state.patch.txt)

Let me know if it works, I will commit it then.

**Comment by jktjkt** [comment]
I looked into this over the weekend, and after rather extensive testing, here's what I've found:

- if the `leaf-list` is system-ordered (which is the default), I just don't know how to use libyang to insert duplicate items. As soon as I insert a duplicate, it gets silently discarded. I only ever see each distinct value once from other sessions -- both in the change iterator and also when requesting data via `Session::get_data`.

- For `ordered-by: user` `leaf-list`, with your patch applied, I eventually hit another error within sysrepo. The gist is that sysrepo starts processing individual items in the diff to be applied, it sees that there's a `lyd_attr` named `yang:value`, and that attribute points to a value which is not present in the new `leaf-list` anymore. In the specific test I used, I tried to go from `[10, 20, 0, -5]` to `[0, 1, 0]`, and when the error happened, the `yang:value` was set to `20`. I think the second `0` was the culprit. I wanted to change the code so that instead of saying "this items is inserted after X" it says "this item is inserted at the last position" -- and that sounds also like turning an 𝒪(n²) operation into 𝒪(n) -- but I got lost in the code and wasn't able to find the place where this thing starts causing trouble, I was only dealing with the fallback.

In the end I gave up. I pondered using YANG extensions and introducing a whole new type into libyang, perhaps as an optimization for a magical `leaf-list` that is always handled as a single thing internally (and which therefore doesn't support plenty of YANG features such as must, ranges, etc), but in the end the simplest way was changing the YANG model and creating an `anydata` object for this purpose. This has a number of disadvantages:

- There's no machine readable description.
- It cannot be validated.
- In the current libyang architecture, I can provide either some pre-baked value suitable for JSON serialization (`LYD_ANYDATA_JSON`), or I can provide a ready-to-use XML string (`LYD_ANYDATA_SXML`). If I provide JSON, then the data cannot be fetched from Netopeer2. If I provide them in the XML format, I cannot call `lyd_print_mem` to get a format suitable for RESTCONF. I'm missing a feature in libyang that would allow me to provide both JSON and XML serializations of the same data at the same time (or perhaps even a callback if you preferred to print-on-demand).

On the other hand, this approach also has advantages:

- The memory churn is substantially reduced because everything is just one item from libyang's point of view.
- It also prevents the code at the client side (which might try to be helpful YANG stack and focus on diff computation even if sysrepo avoided that on the server side) from hitting an 𝒪(n²) code path many times a second on a ~20k vector.
- Interestingly, the fact that the whole `anydata` blob is silently discarded when requested over NETCONF might be considered a feature for us. This way, our [interactive console](https://github.com/CESNET/netconf-cli) (that speaks NETCONF) won't see the data, while the [RESTCONF-ish gateway](https://gerrit.cesnet.cz/c/CzechLight/rousette/+/4005) can simply call `lyd_print_mem' as usual. This saves us adding an annotation and implementing a feature for skipping some nodes in the CLI :). OK, I suspect this should be fixed, but it's not necessarily a problem for us today.

TL;DR: `anydata` looks like a reasonable compromise, especially for non-scalar telemetry data. Sooner than later, however, this will need an implementation in libyang that supports both XML and JSON serializations at the same time. If you have time in future, I also suggest exploring options of (optionally) treating some `leaf-list`s as single entities. Perhaps a special YANG annotation?

**Comment by michalvasko** [comment]
I see, so it seems the patch is not really needed and you have found an acceptable alternate solution. The plan for the future is to properly support these problematic use-cases (state leaf-list, key-less list) but that is still months away. So, sorry about these problems, we simply thought this feature can wait until we have enough time to properly look at it and try to implement it while being aware of all the problems.
